energy-dao/src/external_sc_interactions/farm_config.rs

```diff
@@ -0,0 +1,191 @@
+                farm_token_nonce: 0u64,
+                reward_token_nonce: 0u64,
+                farm_unstaked_value: BigUint::zero(),
+                reward_reserve: BigUint::zero(),
+                farm_rps: BigUint::zero(),
+                division_safety_constant,
+            };
+            farm_config_mapper.set(farm_config);
+        }
+    }
+
+    #[only_owner]
+    #[endpoint(removeFarms)]
+    fn remove_farms(&self, farms: MultiValueEncoded<ManagedAddress>) {
+        for farm in farms {
+            let farm_config_mapper = self.farm_config(&farm);
+            require!(!farm_config_mapper.is_empty(), ERROR_FARM_DOES_NOT_EXIST);
+            let farm_config = farm_config_mapper.get();
+            require!(farm_config.farm_staked_value == 0, ERROR_FARM_HAS_FUNDS);
+            farm_config_mapper.clear();
+        }
+    }
+
+    fn mint_tokens<T: TopEncode>(
+        &self,
+        token_id: TokenIdentifier,
+        amount: BigUint,
+        attributes: &T,
+    ) -> EsdtTokenPayment<Self::Api> {
+        let new_nonce = self
+            .send()
+            .esdt_nft_create_compact(&token_id, &amount, attributes);
+
+        EsdtTokenPayment::new(token_id, new_nonce, amount)
+    }
+
+    fn burn_tokens(&self, token_id: &TokenIdentifier, nonce: Nonce, amount: &BigUint) {
+        self.send().esdt_local_burn(token_id, nonce, amount);
+    }
+
+    fn get_token_attributes<T: TopDecode>(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+    ) -> T {
+        let token_info = self.blockchain().get_esdt_token_data(
+            &self.blockchain().get_sc_address(),
+            token_id,
+            token_nonce,
+        );
+
+        token_info.decode_attributes()
+    }
+
+    #[view(getFarmConfig)]
+    fn get_farm_config(&self, farm_address: &ManagedAddress) -> FarmConfig<Self::Api> {
+        let farm_config_mapper = self.farm_config(farm_address);
+        require!(!farm_config_mapper.is_empty(), "Farm does not exist");
+        farm_config_mapper.get()
+    }
+
+    #[inline]
+    fn get_farming_token(&self, farm_address: &ManagedAddress) -> TokenIdentifier {
+        let farming_token_id = self.farming_token_id().get_from_address(farm_address);
+        self.require_valid_token_id(&farming_token_id);
+        farming_token_id
+    }
+
+    #[inline]
+    fn get_farm_token(&self, farm_address: &ManagedAddress) -> TokenIdentifier {
+        let farm_token_id = self.farm_token_id().get_from_address(farm_address);
+        self.require_valid_token_id(&farm_token_id);
+        farm_token_id
+    }
+
+    #[storage_mapper("farm_token_id")]
+    fn farm_token_id(&self) -> SingleValueMapper<TokenIdentifier>;
+
+    #[view(getFarmingTokenId)]
+    #[storage_mapper("farming_token_id")]
+    fn farming_token_id(&self) -> SingleValueMapper<TokenIdentifier>;
+
+    #[storage_mapper("lockedTokenId")]
+    fn locked_token_id(&self) -> SingleValueMapper<TokenIdentifier>;
```

Comment:
locked token ID can be read through the `get_locked_token_id` in the `EnergyQueryModule`, no need to store it here.

Reply:
Updated.